test/dynamo/test_misc.py

```diff
@@ -4273,6 +4286,129 @@ def dummy_fn():
             # TODO should also pass the code object back into dynamo again, but
             # dynamo is not enabled for Python 3.11 yet.
 
+    @torch._dynamo.config.patch(dynamic_shapes=True)
+    def test_raise_guard_full_constraint(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] == 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.optimize("eager")(my_dyn_fn)(y)
+        torch._dynamo.mark_dynamic(y, 0)
+
+        torch._dynamo.reset()
+        with self.assertRaises(
+            torch._dynamo.exc.InternalTorchDynamoError,
+        ):
+            torch._dynamo.optimize("eager")(my_dyn_fn)(y)
+
+    @torch._dynamo.config.patch(dynamic_shapes=True)
+    def test_no_raise_guard_partial_constraint(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.optimize("eager")(my_dyn_fn)(y)
+        torch._dynamo.mark_dynamic(y, 0)
+        torch._dynamo.reset()
```

Comment:
Hmph. The reset here is a bit irritating. Maybe we should guard on the dynamic markers on the tensors.

Reply:
Oh interesting suggestion - but I kind of like the idea of these being counter to guards - in that they alleviate them. There is a lot of value in keeping them separate, I think. I think the idea here would be that we do not really have the data anymore of what was and was not marked dynamic - for all cases. However, if we want, this would be a very easy guard to add :)

Reply:
My main concern is a benchmarking one: if you warm up the cache with a static compile, then mark a dimension as dynamic and run with the same size, you will hit the static (and not recompile with dynamic shapes).

Reply:
True - Ostensibly you would mark dynamic ahead of time - we could also assert and manually request a .reset() - the way we do if a backend changes w/o a reset. Thoughts?

Reply:
There's no where to assert... except in the guard lol

Reply:
Let me hold off on this, and add a comment, I want to think about this more.